### PR TITLE
Add Hugging Face model download script

### PR DESCRIPTION
**Pull request by jerrychen109**
This script can be used to download Hugging Face model files, either directly from Hugging Face Hub or from a server that hosts the model files in the same structure as the [HF Hub cache system](https://huggingface.co/docs/huggingface_hub/guides/manage-cache)

### CODE REVIEW

File: scripts/misc/download_hf_model.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+SAFETENSORS_WEIGHTS_NAME = "model.safetensors"
+SAFETENSORS_WEIGHTS_INDEX_NAME = "model.safetensors.index.json"
+SAFETENSORS_EXCLUDE_PATTERN = "model*.safetensors*"
+
+DEFAULT_HF_HUB_CACHE_DIR = os.getenv(
+    "HUGGINGFACE_HUB_CACHE", os.path.expanduser("~/.cache/huggingface/hub")
+)
+
+
+def download_from_hf_hub(
```
Comment (dakinggg) [inline scripts/misc/download_hf_model.py:35]: lets move both of these functions into the main llmfoundry library code, and then write unit tests for them.

File: scripts/misc/download_hf_model.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+        "*.ckpt",
+        "*.h5",
+        "*.msgpack",
+    ]
+
+    if (
+        SAFETENSORS_WEIGHTS_NAME in repo_files
+        or SAFETENSORS_WEIGHTS_INDEX_NAME in repo_files
+    ) and prefer_safetensors:
+        print("Safetensors found and preferred. Excluding pytorch files")
```
Comment (dakinggg) [inline scripts/misc/download_hf_model.py:68]: lets use logging instead of prints

File: scripts/misc/download_hf_model.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+):
+    """Downloads all files from a directory on a remote server, including subdirectories.
+
+    Args:
+        session: A requests.Session through which to make requests to the remote server.
+        url (str): The base URL where the files are located.
+        path (str): The path from the base URL to the files to download. The full URL for the download is equal to
+            '<base_url>/<path>'.
+        save_dir (str): The directory to save downloaded files to.
+        ignore_cert (bool): Whether or not to ignore the validity of the SSL certificate of the remote server.
```
Comment (dakinggg) [inline scripts/misc/download_hf_model.py:120]: whats the purpose of even having this as an option?

File: scripts/misc/download_hf_model.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+        PermissionError: If the remote server returns a 401 Unauthorized status code.
+        RuntimeError: If the remote server returns a status code other than 200 OK or 401 Unauthorized.
+    """
+    url = f"{base_url}/{path}"
+    response = session.get(url, verify=(not ignore_cert))
+
+    if response.status_code == HTTPStatus.UNAUTHORIZED:
+        raise PermissionError(
+            f"Could not download file from {url}. Received status code {response.status_code}. "
+            + "Please check that your token is valid and try again."
```
Comment (dakinggg) [inline scripts/misc/download_hf_model.py:133]: what token is this referring to?

File: scripts/misc/download_hf_model.py (hunk 3 above)
Comment (jerrychen109) [reply scripts/misc/download_hf_model.py:120]: If you want to host your own cache server with a self-signed TLS certificate, you'll need to specify this option

File: scripts/misc/download_hf_model.py (hunk 4 above)
Comment (jerrychen109) [reply scripts/misc/download_hf_model.py:133]: Hugging Face Hub API token - will make clarify

File: scripts/misc/download_hf_model.py (hunk 1 above)
Comment (jerrychen109) [reply scripts/misc/download_hf_model.py]: Moved into `llmfoundry/utils/model_download_utils.py`. Unit testing is a little tricky though - not sure if there's a clean way to test file downloads over network. I could probably use `pyfakefs` somehow, but I'm not sure how to easily mock the Hugging Face or cache server behavior

Comment (dakinggg) [reply scripts/misc/download_hf_model.py]: i'd say don't worry about testing the network call itself, so just mock `snapshot_download` and `session.get`, to just test the code around the network call.

File: scripts/misc/download_hf_model.py [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+        default=False,
+        help='Whether to fallback to downloading from Hugging Face if download from cache fails',
+    )
+
+    args = argparser.parse_args(sys.argv[1:])
+    if args.download_from == 'hf':
+        download_from_hf_hub(args.model, save_dir=args.save_dir, token=args.token)
+    else:
+        try:
+            download_from_cache_server(
```
Comment (jerrychen109) [inline scripts/misc/download_hf_model.py:44]: Going to add some retry logic on each call as well

**Review by dakinggg — APPROVED** [review]

**Pull request merged**